### ISSUE TITLE
a small question about batch in gcn_lib/sparse/torch_edge.py

### ISSUE THREAD

**Issue by WMF1997** [lightaime/deep_gcns_torch#5]
hello @guochengqian 
I want to check what the layers is going on in ```gcn_lib/sparse/torch_edge.py```. And I just wonder the line 
```batch_size=batch[-1]+1```, so... what is ```batch``` argument? i.e., ```type(batch)```, or more specifically ```batch.shape``` should be what? (I guess it is a ```torch.Tensor````? )

here is part of the code I want to ask, with Chinese comment: 
```python
def knn_matrix(x, k=16, batch=None):
    """Get KNN based on the pairwise distance.
    Args:
        pairwise distance: (num_points, num_points)
        k: int
    Returns:
        nearest neighbors: (num_points*k ,1) (num_points, k)
    """
    # 这一步骤的操作有点没有想明白, 是不是某个地方写错了?
    batch_size = batch[-1] + 1
    # 转变形状. 
    x = x.view(batch_size, -1, x.shape[-1])

    # 这里修改为负数, 是为什么? 这是因为topk是选择其中最大的数据, 所以为了找到最小的, 自然要添加负数来进行修正. 
    neg_adj = - pairwise_distance(x)
    _, nn_idx = torch.topk(neg_adj, k=k)
    del neg_adj

    # 一共多少个点. 
    n_points = x.shape[1]

    # 加上了起始点的索引之后, 相当于是将原先的情况进行了一个修正. 将上文的相对索引修正成为绝对索引. 
    start_idx = torch.arange(0, n_points*batch_size, n_points).long().view(batch_size, 1, 1)
    if x.is_cuda:
        start_idx = start_idx.cuda()
    nn_idx += start_idx
    del start_idx


    if x.is_cuda:
        torch.cuda.empty_cache()

    nn_idx = nn_idx.view(1, -1)
    # 中心点的坐标
    center_idx = torch.arange(0, n_points*batch_size).repeat(k, 1).transpose(1, 0).contiguous().view(1, -1)
    if x.is_cuda:
        center_idx = center_idx.cuda()
    return nn_idx, center_idx
```

and I add a "main" function to test if it works. (just after the ```gcn_lib/sparse/torch_edge.py```)
```python
if __name__ == "__main__":
    import torch
    torch.manual_seed(0)
    edge_index = torch.randint(0, 10, [2, 40])
    d = Dilated(k=5, dilation=2, stochastic=True, epsilon=0.9)
    e_1 = d(edge_index)

    x = torch.randn([10, 20])
    batch = torch.randint(0, 3, [10, ])
    dknn = DilatedKnnGraph(k=5)
    e_2 = dknn(x, batch)
```
here is the problem:
![image](https://user-images.githubusercontent.com/37621591/65400745-8e439d80-ddf6-11e9-821d-9ca2ca3f9eea.png)

when i changed the seed from 0 to 1, then everything goes fine?

![image](https://user-images.githubusercontent.com/37621591/65400785-ab786c00-ddf6-11e9-9163-18fd30519d55.png)

yours sincerely, 
@wmf1997

**Comment by guochengqian** [comment]
Dear WMF1997, 

In your example, the batch variable is wrong. change to `batch = torch.randint(0, 0, [10, ])` it will work.  This means every node blong to the same batch, batch 0. 
What's batch?
This code is based on Pytorch_geometric. I highly recommend you to read introduction of pytorch geometirc at first.  
Simply speacking, for graph, the data shape is (N, C), N is the sum of all the graph nodes in one mini-batch. Because every graph may have varied nodes number, so you can not use (B, C, N, 1) data shape to represent graph data with varied nodes in mini-batch way. Batch variable is a vector, whose shape is (N, 1), which indicates that each node belongs to which batch. See this for details. https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html?highlight=batch#mini-batches <https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html?highlight=batch#mini-batches>


With regards,
Gordon

**Comment by WMF1997** [comment]
Thank you very much~ I understand.
